### PR TITLE
Create EventCard Component

### PR DESCRIPTION
**Pull request by alexzhang1618**
# Changes

- Fixed some sizing bugs in Typography (small was larger than large, ...)
- Added CommunityLogo in components to render the logo for a community
- Added an event display on the dashboard along with event cards that pop up modals when clicked
  - Components for this are under components/event
- Added the `ics` package to support calendar exports for events
- Add a new API call to get attendances for the current user

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) 

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots
Screenshot 2023-12-20 at 3 45 41 PM
Screenshot 2023-12-20 at 3 46 00 PM
Screenshot 2023-12-20 at 3 46 12 PM

### CODE REVIEW

File: src/components/events/CalendarButtons/index.tsx
Comment (SheepTester) [inline src/components/events/CalendarButtons/index.tsx]: Should these be links? This way, users can see that they'll be sent to Google Calendar, or they can copy the URL or share the link to an app or another device

The "Add to Apple Calendar" button could also be a download link as well since `createDownloadFile` also creates an `<a>` element

File: src/components/events/CalendarButtons/style.module.scss
Comment (SheepTester) [inline src/components/events/CalendarButtons/style.module.scss]: Maybe it should be left aligned so the icons and "Add to" lies up?

```suggestion
  justify-content: flex-start;
```

though for some reason the icons don't seem to be properly aligned (the Google logo is off center)

image

File: src/components/events/EventCard/index.tsx
Comment (SheepTester) [inline src/components/events/EventCard/index.tsx]: this could be simplified to
```suggestion
  let displayCover = cover || '/assets/graphics/store/hero-photo.jpg';
```

File: src/components/events/EventCard/style.module.scss
Comment (SheepTester) [inline src/components/events/EventCard/style.module.scss]: what is this used for

File: src/components/events/EventCard/style.module.scss
Comment (SheepTester) [inline src/components/events/EventCard/style.module.scss]: Is there a reason why a scrollbar would appear?

File: src/components/events/EventCard/style.module.scss
Comment (SheepTester) [inline src/components/events/EventCard/style.module.scss]: The ellipses don't show for me

image

It seems you have to set these two CSS properties directly on the `Typography` elements themselves in order for it to work

File: src/components/events/CalendarButtons/style.module.scss
Comment (SheepTester) [inline src/components/events/CalendarButtons/style.module.scss]: you should move this `transition` property to `.calendarButton` itself. this way, there's also an animation when your mouse moves off the button

Also, you could move this rule inside `.calendarButton` like this:

```scss
.calendarButton {
  ...
  transition: all 0.2s;
  
  &:hover {
    background: var(--theme-primary-2);
  }
}
```

This has the same result, but it makes it more clear that these two rules are related

File: src/components/events/EventCard/style.module.scss
Comment (SheepTester) [inline src/components/events/EventCard/style.module.scss]: you should also move this `transition` property to `.container` so there's an animation when you move your mouse off the event card

File: src/components/events/EventModal/index.tsx
Comment (SheepTester) [inline src/components/events/EventModal/index.tsx]:
```suggestion
  let displayCover = cover || '/assets/graphics/store/hero-photo.jpg';

  let displayEventLink = eventLink || `https://acmucsd.com/events/${event.uuid}`;
```

you could also use `??` instead of `||` if you're feeling fancy

File: src/components/events/EventModal/index.tsx
Comment (SheepTester) [inline src/components/events/EventModal/index.tsx:1]: In #91 I'm adding a generic `Modal` component so there's less code duplication between this and `HelpModal`. I can migrate this component to use `Modal` if this PR gets merged first

File: src/components/events/EventModal/style.module.scss
Comment (SheepTester) [inline src/components/events/EventModal/style.module.scss:63]: the modal is very narrow on mobile. this is because of this `width: 70vw` line, but you could also remove the horizontal padding on `.modal` when the screen is narrow enough, so there's more horizontal screen space to work with

current | suggested
--- | ---
image | image

File: src/components/events/EventModal/style.module.scss
Comment (SheepTester) [inline src/components/events/EventModal/style.module.scss:65]: image

you should add some gap between the title and the calendar buttons

File: src/components/events/PointsDisplay/index.tsx
Comment (SheepTester) [inline src/components/events/PointsDisplay/index.tsx]: I'd explore alternate ways to show whether you've attended an event (unless we want to keep it hidden?). Green vs blue isn't exactly clear, and relying on color isn't great for colorblind people. One idea is to show a check icon if you've attended an event

File: src/components/events/EventModal/index.tsx
Comment (SheepTester) [reply src/components/events/EventModal/index.tsx]: Should this also use `formatURLEventTitle` to include the event title in the URL, like what is used for the Add to Apple Calendar button?

**Review by SheepTester — CHANGES_REQUESTED** [review]

File: src/components/common/CommunityLogo/index.tsx
Comment (farisashai) [inline src/components/common/CommunityLogo/index.tsx]: nit: I would rename this to size if it's going to define both the width and height as the same value

File: src/components/events/EventModal/index.tsx
Comment (farisashai) [inline src/components/events/EventModal/index.tsx]:
```suggestion
import { CommunityLogo } from '@/components/common';
```
super tiny nit: try to import from the enclosing folder to ensure imports are always named correctly

File: src/components/events/EventModal/style.module.scss
Comment (farisashai) [inline src/components/events/EventModal/style.module.scss]: nit: don't hardcode breakpoints like this otherwise things get really messy really fast - there are preset breakpoints that are standard conventions defined in the global variables vars.scss

$breakpoint-md: 768px; // Use this breakpoint for most mobile-first CSS designs

File: src/components/events/EventModal/style.module.scss
Comment (farisashai) [inline src/components/events/EventModal/style.module.scss:19]: this seems like the start of a very difficult z-index system to maintain - possibly consider any other approaches

File: src/components/events/PointsDisplay/style.module.scss
Comment (farisashai) [inline src/components/events/PointsDisplay/style.module.scss]:
```suggestion
  background-color: var(--theme-primary-4);
```


you don't need a class name for blue since this is essentially the default background color for all .circle elements - i would only keep the green one as a more specific override but also rename it for specificity to attended or something better. i agree w sean that we shouldn't be using colors to indicate attendance anyways

File: src/components/events/EventModal/style.module.scss
Comment (farisashai) [inline src/components/events/EventModal/style.module.scss:96]: what's the reason for the !important? it's always never necessary if you specific class names and elements correctly so i would try to remove it and solve this differently

**Review by farisashai — CHANGES_REQUESTED** [review]
sean covered everything visual mostly - biggest feedback is reuse his modal component or find a way to unify the two and resolve the scrolling underneath the modal bug

File: src/components/events/CalendarButtons/index.tsx
Comment (alexzhang1618) [reply src/components/events/CalendarButtons/index.tsx]: 👍

File: src/components/events/CalendarButtons/style.module.scss
Comment (alexzhang1618) [reply src/components/events/CalendarButtons/style.module.scss]: I've updated the logos since there was a weird gap in the Google logo, and fixed the alignment

File: src/components/events/EventCard/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventCard/style.module.scss]: I think it's leftover CSS, deleted 😓

File: src/components/events/EventCard/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventCard/style.module.scss]: removed 👍

File: src/components/events/EventCard/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventCard/style.module.scss]: fixed o7

File: src/components/events/CalendarButtons/style.module.scss
Comment (alexzhang1618) [reply src/components/events/CalendarButtons/style.module.scss]: good catch! fixed

File: src/components/events/EventModal/index.tsx
Comment (alexzhang1618) [reply src/components/events/EventModal/index.tsx]: hm `displayEventLink` just links to a main website preview of the event, so we need to keep it this way by convention, so probably not

File: src/components/events/EventModal/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventModal/style.module.scss:63]: fixed !

File: src/components/events/EventModal/index.tsx
Comment (alexzhang1618) [reply src/components/events/EventModal/index.tsx:1]: Sure, let's revisit this after this PR and #91 are merged and done

File: src/components/events/EventModal/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventModal/style.module.scss:65]: done!

File: src/components/events/PointsDisplay/index.tsx
Comment (alexzhang1618) [reply src/components/events/PointsDisplay/index.tsx]: I think we should iterate a little with design before settling here, I like the idea of a check but it does make the point total unreadable 🤔

Comment (alexzhang1618) [reply src/components/events/PointsDisplay/index.tsx]: Maybe for now we can keep green and update in a future PR once we've figured out a direction on this

File: src/components/events/EventModal/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventModal/style.module.scss]: good catch, I didn't know this existed 😮

File: src/components/events/EventModal/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventModal/style.module.scss:96]: So for the `<Typography>` component, it specifies its own styles for the text, so I need to add `!important` when trying to downsize for mobile views, or else it won't update.

I could try to update the variant for `Typography` instead of using CSS here, but I think this is a little simpler. I'll add some comments here also.

File: src/components/events/EventModal/style.module.scss
Comment (alexzhang1618) [reply src/components/events/EventModal/style.module.scss:19]: @SheepTester maybe some food for thought 😁

**Review by farisashai — APPROVED** [review]
looks fine at a quick glance, approving so it's not blocked - sean can do a sanity check if he wants

File: src/components/events/EventModal/style.module.scss
Comment (SheepTester) [reply src/components/events/EventModal/style.module.scss:19]: we could store z-indices as constants in vars.scss

File: src/components/events/CalendarButtons/index.tsx
Comment (SheepTester) [inline src/components/events/CalendarButtons/index.tsx:65]: you should probably put this in a `useMemo` or something. it's probably not a big deal because the ics file probably isn't that big, and there aren't that many interactive elements on the page, but every time React re-renders, it will create a new object url, which would lead to a memory leak

Something like

```ts
const url = useMemo(() => {
  const file = ...;
  return URL.createObjectURL(file);
}, [...]);

useEffect(() => {
  return () => {
    URL.revokeObjectURL(url);
  };
}, [url]);
```

maybe

**Review by SheepTester — APPROVED** [review]
lgtm

File: src/components/events/CalendarButtons/index.tsx
Comment (alexzhang1618) [reply src/components/events/CalendarButtons/index.tsx:65]: I've been playing around with this for like an hour and for some reason this approach isn't really working.

It seems like the blob/objectURL no longer works after refreshing the page, so after hitting refresh none of the apple calendar buttons work.

I'm unsure if this is a local-only problem because the console error after hitting the apple calendar link after a refresh is `Not allowed to load local resource: blob:nodedata:...`

I'm going to push what I have sans the `useMemo` for appleCal and we can revisit in the future

**Pull request merged**
